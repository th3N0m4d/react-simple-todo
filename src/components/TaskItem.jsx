import React from 'react'
import PropTypes from 'prop-types'
import Button from 'react-bootstrap/Button'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCheck, faTrash, faUndo } from '@fortawesome/free-solid-svg-icons'
import cx from 'classnames'

const propTypes = {
  author: PropTypes.string,
  name: PropTypes.string,
  onConfirm: PropTypes.func,
  onRemove: PropTypes.func,
  onToggle: PropTypes.func
}

const defaultProps = {
  name: 'N/A',
  author: 'Anonymous',
  completed: false
}

const renderConfirmButton = onClick => (
  <Button variant='outline-success' onClick={onClick}>
    <FontAwesomeIcon icon={faCheck} />
  </Button>
)

const renderUndoButton = onClick => (
  <Button variant='outline-info' onClick={onClick}>
    <FontAwesomeIcon icon={faUndo} />
  </Button>
)

const TaskItem = ({
  author,
  name,
  onConfirm,
  onRemove,
  completed,
  onToggle
}) => (
  <li className='list-group-item'>
    <div className='todo-indicator bg-warning' />
    <div className='widget-content p-0'>
      <div className='widget-content-wrapper'>
        <div className='widget-content-left mr-2'>
          <div className='custom-checkbox custom-control'>
            <input className='custom-control-input' id='exampleCustomCheckbox12' type='checkbox' />
            <label className='custom-control-label' htmlFor='exampleCustomCheckbox12'>&nbsp;</label>
          </div>
        </div>
        <div className='widget-content-left'>
          <div className={cx('widget-heading', {
            'widget-heading--completed': completed
          })}
          >
            {name}
          </div>
          <div className='widget-subheading'><i>By {author}</i></div>
        </div>
        <div className='widget-content-right'>

          {
            completed
              ? renderUndoButton(onToggle)
              : renderConfirmButton(onToggle)
          }
          <Button variant='outline-danger' onClick={onRemove}>
            <FontAwesomeIcon icon={faTrash} />
          </Button>

        </div>
      </div>
    </div>
  </li>
)

TaskItem.propTypes = propTypes
TaskItem.defaultProps = defaultProps

export default TaskItem
