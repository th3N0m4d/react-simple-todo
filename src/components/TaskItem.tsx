import * as React from 'react'
import Button from 'react-bootstrap/Button'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCheck, faTrash, faUndo } from '@fortawesome/free-solid-svg-icons'
import cx from 'classnames'
import Variants from '@/constants/Variants'

interface Props {
  dueDate?: number,
  completed?: boolean,
  name?: string,
  onRemove?: () => void,
  onToggle?: () => void,
}

const renderConfirmButton = (onClick: any) => (
  <Button variant='outline-success' onClick={onClick} data-test="confirm-button">
    <FontAwesomeIcon icon={faCheck} />
  </Button>
)

const renderUndoButton = (onClick: any) => (
  <Button variant='outline-info' onClick={onClick} data-test="undo-button">
    <FontAwesomeIcon icon={faUndo} />
  </Button>
)

const dateFormatOptions = { weekday: 'long', month: 'long', day: 'numeric' }

const getVariant = (time: number | undefined) => {
  if (time) {
    const now = new Date().getTime()

    if (now > time) {
      return Variants.danger
    }
  }

  return Variants.success
}

const TaskItem: React.FunctionComponent<Props> = ({
  dueDate,
  completed,
  name,
  onRemove,
  onToggle,
}) => (
    <>
      <div
        className={cx('todo-indicator', {
          [`bg-${getVariant(dueDate)}`]: true
        })}
      />
      <div className='widget-content p-0'>
        <div className='widget-content-wrapper'>
          <div className='widget-content-left'>
            <div className={cx('widget-heading', {
              'widget-heading--completed': completed
            })}
              data-test="heading"
            >
              {name}
            </div>
            {dueDate && <div className='widget-subheading' data-test="subheading"><i>Due date: {new Intl.DateTimeFormat('en-US', dateFormatOptions).format(dueDate)}</i></div>}
          </div>
          <div className='widget-content-right'>

            {
              completed
                ? renderUndoButton(onToggle)
                : renderConfirmButton(onToggle)
            }
            <Button variant='outline-danger' onClick={onRemove}>
              <FontAwesomeIcon icon={faTrash} />
            </Button>

          </div>
        </div>
      </div>
    </>
  )


export default TaskItem
