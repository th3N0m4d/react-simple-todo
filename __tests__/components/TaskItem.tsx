import * as React from 'react'
import { shallow } from 'enzyme'
import Button from 'react-bootstrap/Button'

import TaskItem from '@/components/TaskItem'

describe('TaskItem', () => {
  describe('Props', () => {
    it('should render component correctly', () => {
      const wrapper = shallow(<TaskItem />)

      const subheading = wrapper.find('[data-test="subheading"]')
      const confirmButton = wrapper.find('[data-test="confirm-button"]')
      const undoButton = wrapper.find('[data-test="undo-button"]')
      const heading = wrapper.find('[data-test="heading"]')

      expect(wrapper).toMatchSnapshot()
      expect(subheading.exists()).toBeFalsy()
      expect(confirmButton.exists()).toBeTruthy()
      expect(undoButton.exists()).toBeFalsy()
      expect(heading.hasClass('widget-heading--completed')).toBeFalsy()
    })

    it.each`
      propName        | value             
      ${'dueDate'}    | ${546300000000}   
      ${'name'}       | ${'Buy wine'}     
      ${'completed'}  | ${true}            
    `('should render component with prop $propName and value $value',
      ({ propName, value }) => {
        const props = { [propName]: value }
        const wrapper = shallow(<TaskItem {...props} />)

        expect(wrapper).toMatchSnapshot()
      });

  });

  describe('Interaction', () => {
    it('should trigger onRemove callback', () => {
      const onRemoveSpy = jest.fn()
      const INDEX = 1
      const wrapper = shallow(
        <TaskItem onRemove={onRemoveSpy} />
      )

      const button = wrapper.find(Button).at(INDEX)

      button.simulate('click')

      expect(onRemoveSpy).toHaveBeenCalled()
    })

    it('should trigger onToggle callback for confirm operation', () => {
      const onToggleSpy = jest.fn()
      const INDEX = 0
      const wrapper = shallow(
        <TaskItem onToggle={onToggleSpy} />
      )

      const button = wrapper.find(Button).at(INDEX)

      button.simulate('click')

      expect(onToggleSpy).toHaveBeenCalled()
    })

    it('should trigger onToggle callback for undo operation', () => {
      const onToggleSpy = jest.fn()
      const INDEX = 0
      const wrapper = shallow(
        <TaskItem onToggle={onToggleSpy} completed={true} />
      )

      const button = wrapper.find(Button).at(INDEX)

      button.simulate('click')

      expect(onToggleSpy).toHaveBeenCalled()
    })
  });

})
