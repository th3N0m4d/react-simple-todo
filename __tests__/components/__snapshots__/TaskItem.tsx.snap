// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskItem Props should render component correctly 1`] = `
<Fragment>
  <div
    className="todo-indicator bg-success"
  />
  <div
    className="widget-content p-0"
  >
    <div
      className="widget-content-wrapper"
    >
      <div
        className="widget-content-left"
      >
        <div
          className="widget-heading"
          data-test="heading"
        />
      </div>
      <div
        className="widget-content-right"
      >
        <Button
          active={false}
          data-test="confirm-button"
          disabled={false}
          type="button"
          variant="outline-success"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  512,
                  512,
                  Array [],
                  "f00c",
                  "M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z",
                ],
                "iconName": "check",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </Button>
        <Button
          active={false}
          disabled={false}
          type="button"
          variant="outline-danger"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  448,
                  512,
                  Array [],
                  "f1f8",
                  "M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z",
                ],
                "iconName": "trash",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </Button>
      </div>
    </div>
  </div>
</Fragment>
`;

exports[`TaskItem Props should render component with prop completed and value true 1`] = `
<Fragment>
  <div
    className="todo-indicator bg-success"
  />
  <div
    className="widget-content p-0"
  >
    <div
      className="widget-content-wrapper"
    >
      <div
        className="widget-content-left"
      >
        <div
          className="widget-heading widget-heading--completed"
          data-test="heading"
        />
      </div>
      <div
        className="widget-content-right"
      >
        <Button
          active={false}
          data-test="undo-button"
          disabled={false}
          type="button"
          variant="outline-info"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  512,
                  512,
                  Array [],
                  "f0e2",
                  "M212.333 224.333H12c-6.627 0-12-5.373-12-12V12C0 5.373 5.373 0 12 0h48c6.627 0 12 5.373 12 12v78.112C117.773 39.279 184.26 7.47 258.175 8.007c136.906.994 246.448 111.623 246.157 248.532C504.041 393.258 393.12 504 256.333 504c-64.089 0-122.496-24.313-166.51-64.215-5.099-4.622-5.334-12.554-.467-17.42l33.967-33.967c4.474-4.474 11.662-4.717 16.401-.525C170.76 415.336 211.58 432 256.333 432c97.268 0 176-78.716 176-176 0-97.267-78.716-176-176-176-58.496 0-110.28 28.476-142.274 72.333h98.274c6.627 0 12 5.373 12 12v48c0 6.627-5.373 12-12 12z",
                ],
                "iconName": "undo",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </Button>
        <Button
          active={false}
          disabled={false}
          type="button"
          variant="outline-danger"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  448,
                  512,
                  Array [],
                  "f1f8",
                  "M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z",
                ],
                "iconName": "trash",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </Button>
      </div>
    </div>
  </div>
</Fragment>
`;

exports[`TaskItem Props should render component with prop dueDate and value 546300000000 1`] = `
<Fragment>
  <div
    className="todo-indicator bg-danger"
  />
  <div
    className="widget-content p-0"
  >
    <div
      className="widget-content-wrapper"
    >
      <div
        className="widget-content-left"
      >
        <div
          className="widget-heading"
          data-test="heading"
        />
        <div
          className="widget-subheading"
          data-test="subheading"
        >
          <i>
            Due date: 
            Saturday, April 25
          </i>
        </div>
      </div>
      <div
        className="widget-content-right"
      >
        <Button
          active={false}
          data-test="confirm-button"
          disabled={false}
          type="button"
          variant="outline-success"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  512,
                  512,
                  Array [],
                  "f00c",
                  "M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z",
                ],
                "iconName": "check",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </Button>
        <Button
          active={false}
          disabled={false}
          type="button"
          variant="outline-danger"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  448,
                  512,
                  Array [],
                  "f1f8",
                  "M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z",
                ],
                "iconName": "trash",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </Button>
      </div>
    </div>
  </div>
</Fragment>
`;

exports[`TaskItem Props should render component with prop name and value Buy wine 1`] = `
<Fragment>
  <div
    className="todo-indicator bg-success"
  />
  <div
    className="widget-content p-0"
  >
    <div
      className="widget-content-wrapper"
    >
      <div
        className="widget-content-left"
      >
        <div
          className="widget-heading"
          data-test="heading"
        >
          Buy wine
        </div>
      </div>
      <div
        className="widget-content-right"
      >
        <Button
          active={false}
          data-test="confirm-button"
          disabled={false}
          type="button"
          variant="outline-success"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  512,
                  512,
                  Array [],
                  "f00c",
                  "M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z",
                ],
                "iconName": "check",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </Button>
        <Button
          active={false}
          disabled={false}
          type="button"
          variant="outline-danger"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  448,
                  512,
                  Array [],
                  "f1f8",
                  "M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z",
                ],
                "iconName": "trash",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </Button>
      </div>
    </div>
  </div>
</Fragment>
`;
